storage/cluster/database.go

```diff
@@ -0,0 +1,199 @@
+	newStorageDatabase newStorageDatabaseFn = storage.NewDatabase
+
+	errAlreadyWatchingTopology = errors.New("cluster database is already watching topology")
+	errNotWatchingTopology     = errors.New("cluster database is already watching topology")
+)
+
+type newStorageDatabaseFn func(
+	namespaces []namespace.Metadata,
+	shardSet sharding.ShardSet,
+	opts storage.Options,
+) (storage.Database, error)
+
+type clusterDB struct {
+	storage.Database
+
+	log    xlog.Logger
+	hostID string
+	watch  topology.MapWatch
+
+	watchMutex sync.Mutex
+	watching   bool
+	doneCh     chan struct{}
+	closedCh   chan struct{}
+}
+
+// NewDatabase creates a new clustered time series database
+func NewDatabase(
+	namespaces []namespace.Metadata,
+	hostID string,
+	topoInit topology.Initializer,
+	opts storage.Options,
+) (Database, error) {
+	log := opts.InstrumentOptions().Logger()
+	topo, err := topoInit.Init()
+	if err != nil {
+		return nil, err
+	}
+
+	watch, err := topo.Watch()
+	if err != nil {
+		return nil, err
+	}
+
+	// Wait for the topology to be available
+	<-watch.C()
+
+	shardSet, ok := hostOrEmptyShardSet(hostID, watch.Get())
+	if !ok {
+		log.Warnf("topology has no shard set for host ID: %s", hostID)
+	}
+
+	db, err := newStorageDatabase(namespaces, shardSet, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	return &clusterDB{
+		Database: db,
+		log:      log,
+		hostID:   hostID,
+		watch:    watch,
+	}, nil
+}
+
+func (d *clusterDB) Open() error {
+	select {
+	case <-d.watch.C():
+		shardSet, ok := hostOrEmptyShardSet(d.hostID, d.watch.Get())
+		if !ok {
+			d.log.Warnf("topology has no shard set for host ID: %s", d.hostID)
+		}
+		d.Database.AssignShardSet(shardSet)
+	default:
+		// No updates to the topology since cluster DB created
+	}
+
+	if err := d.Database.Open(); err != nil {
+		return err
+	}
+	if err := d.startActiveTopologyWatch(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (d *clusterDB) Close() error {
+	if err := d.Database.Close(); err != nil {
+		return err
+	}
+	if err := d.stopActiveTopologyWatch(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (d *clusterDB) startActiveTopologyWatch() error {
+	d.watchMutex.Lock()
+	defer d.watchMutex.Unlock()
+
+	if d.watching {
+		return errAlreadyWatchingTopology
+	}
+
+	d.watching = true
+
+	d.doneCh = make(chan struct{}, 1)
+	d.closedCh = make(chan struct{}, 1)
+
+	go d.activeTopologyWatch()
+
+	return nil
+}
+
+func (d *clusterDB) stopActiveTopologyWatch() error {
+	d.watchMutex.Lock()
+	defer d.watchMutex.Unlock()
+
+	if !d.watching {
+		return errNotWatchingTopology
+	}
+
+	d.watching = false
+
+	d.doneCh <- struct{}{}
+	close(d.doneCh)
+	<-d.closedCh
+	close(d.closedCh)
+
+	return nil
+}
+
+func (d *clusterDB) activeTopologyWatch() {
+	for {
+		select {
+		case <-d.doneCh:
+			d.closedCh <- struct{}{}
+			return
+		case <-d.watch.C():
+			shardSet, ok := hostOrEmptyShardSet(d.hostID, d.watch.Get())
+			if !ok {
```

Comment:
Yeah it's meant for when you want to deploy new nodes and make them bootstrap as soon as you do a shard reassignment, rather than having to force you to do the reassignment first and make it go into a fatal/crash loop until it has a shard assignment.